kernel/src/fs/cgroupfs/controller/mod.rs

```diff
@@ -0,0 +1,413 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{string::String, sync::Arc, vec::Vec};
+
+use aster_systree::{Error, Result, SysAttrSet, SysAttrSetBuilder};
+use bitflags::bitflags;
+use ostd::{
+    mm::{VmReader, VmWriter},
+    sync::{Mutex, MutexGuard, Rcu},
+    task::{atomic_mode::AsAtomicModeGuard, disable_preempt},
+};
+
+use crate::fs::cgroupfs::{
+    controller::{cpuset::CpuSetController, memory::MemoryController, pids::PidsController},
+    systree_node::CgroupSysNode,
+    CgroupNode,
+};
+
+mod cpuset;
+mod memory;
+mod pids;
+
+/// A trait to abstract all individual cgroup controllers.
+trait SubControl {
+    fn attr_set(&self) -> &SysAttrSet;
+
+    fn read_attr_at(
+        &self,
+        name: &str,
+        offset: usize,
+        writer: &mut VmWriter,
+        cgroup_node: &dyn CgroupSysNode,
+    ) -> Result<usize>;
+
+    fn write_attr(
+        &self,
+        name: &str,
+        reader: &mut VmReader,
+        cgroup_node: &dyn CgroupSysNode,
+    ) -> Result<usize>;
+}
+
+/// Defines the static properties and behaviors of a specific cgroup sub-controller.
+trait SubControlStatic: SubControl + Sized + 'static {
+    /// Creates a new instance of the sub-controller.
+    fn new(is_root: bool) -> Self;
+
+    /// Returns the `ControllerType` enum variant corresponding to this sub-controller.
+    fn controller_type() -> ControllerType;
+
+    /// Obtains the specific `SubController` field from the main `Controller` struct.
+    fn obtain_sub_controller_field(controller: &Controller) -> Arc<SubController<Self>>;
+}
+
+/// The type of sub-controller in the cgroup subsystem.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) enum ControllerType {
```

Comment:
So given that we have `SubCtrlState`. This type should be named as `SubCtrlType`?